tests: gpio driver fails in pin_get_config

**Describe the bug**
When running the testcase  tests/drivers/gpio/gpio_basic_api/ on various stm32 boards, the sequence pin_get_config fails because
flag set differs from flag get. 

At least  on the following target boards : stm32f3_disco,  nucleo_g474re, nucleo_g071rb,  nucleo l152r, nucleo_wl55jc  

During the gpio_stm32_config, the _flags_ is set with value  = 0xa0000                                                                                                                                    
During the gpio_stm32_get_config, the _flags_ is get at value = 0x20000b74 

Moreover it seems that :
- the `gpio_stm32_pincfg_to_flags` and `gpio_stm32_flags_to_conf`  functions have parameters that do not match exactly the 32bit values. (gpio_flags_t is of type uint32_t)
- LL_GPIO_GetPinPull, LL_GPIO_GetPinOutputType, LL_GPIO_GetPinMode  returned values are uint32_t assigned to unsigned int mode, otype, pupd
- When disabling the CONFIG_GPIO_GET_CONFIG (with CONFIG_GPIO_GET_CONFIG=n),  a compilation error occurs:
```
in function `gpio_pin_get_config':
/local/home/frq09495/zephyrproject/zephyr/build/zephyr/include/generated/syscalls/gpio.h:111: undefined reference to `z_impl_gpio_pin_get_config'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**To Reproduce**
Steps to reproduce the behavior:
1. west build -p auto  -b stm32f3_disco tests/drivers/gpio/gpio_basic_api
3. west flash
4. See error

**Expected behavior**
test PASSED 

**Impact**
test campaign failure

**Logs and console output**
```
*** Booting Zephyr OS build zephyr-v3.1.0-2346-g5b5ccee821cf  ***
Running TESTSUITE gpio_basic_test
===================================================================
START - test_gpio_port
Validate device GPIOC
Check GPIOC output 6 connected to input 7
OUT 6 to IN 7 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_logic_output_levels                                                                                                                     
- check_input_levels                                                                                                                            
- bits_logical                                                                                                                                  
- check_pulls                                                                                                                                   
                                                                                                                                                
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:652: pin_get_config: (flags_set not equal to )
flags are different                                                                                                                             
 FAIL - test_gpio_port in 0.37 seconds
```


**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows) ubuntu
 - Toolchain (e.g Zephyr SDK, ...) : zephyr-sdk-0.14.0
 - Commit SHA : 60271524087bb4382c8b4db0f2e17a87e4120065

**Additional context**
PR : https://github.com/zephyrproject-rtos/zephyr/pull/43225

--- a/drivers/gpio/gpio_stm32.c
+++ b/drivers/gpio/gpio_stm32.c
@@ -91,7 +91,7 @@ static int gpio_stm32_flags_to_conf(gpio_flags_t flags, int *pincfg)
 	return 0;
 }
 
-#ifdef CONFIG_GPIO_GET_CONFIG
+#if defined(CONFIG_GPIO_GET_CONFIG) && !defined(CONFIG_SOC_SERIES_STM32F1X)
 /**
  * @brief Custom stm32 flags to zephyr
  */
@@ -563,7 +563,7 @@ static int gpio_stm32_config(const struct device *dev,
 	return 0;
 }
 
-#ifdef CONFIG_GPIO_GET_CONFIG
+#if defined(CONFIG_GPIO_GET_CONFIG) && !defined(CONFIG_SOC_SERIES_STM32F1X)
 /**
  * @brief Get configuration of pin
  */
@@ -657,7 +657,7 @@ static int gpio_stm32_manage_callback(const struct device *dev,
 
 static const struct gpio_driver_api gpio_stm32_driver = {
 	.pin_configure = gpio_stm32_config,
-#ifdef CONFIG_GPIO_GET_CONFIG
+#if defined(CONFIG_GPIO_GET_CONFIG) && !defined(CONFIG_SOC_SERIES_STM32F1X)
 	.pin_get_config = gpio_stm32_get_config,
 #endif /* CONFIG_GPIO_GET_CONFIG */
 	.port_get_raw = gpio_stm32_port_get_raw,

--- a/drivers/gpio/gpio_stm32.c
+++ b/drivers/gpio/gpio_stm32.c
@@ -46,7 +46,7 @@ static void gpio_stm32_isr(int line, void *arg)
 /**
  * @brief Common gpio flags to custom flags
  */
-static int gpio_stm32_flags_to_conf(int flags, int *pincfg)
+static int gpio_stm32_flags_to_conf(gpio_flags_t flags, int *pincfg)
 {
 
 	if ((flags & GPIO_OUTPUT) != 0) {
@@ -91,14 +91,14 @@ static int gpio_stm32_flags_to_conf(int flags, int *pincfg)
 	return 0;
 }
 
+#ifdef CONFIG_GPIO_GET_CONFIG
 /**
  * @brief Custom stm32 flags to zephyr
  */
-#ifdef CONFIG_GPIO_GET_CONFIG
-static int gpio_stm32_pincfg_to_flags(int otype, int pupd, int mode,
-				      int *out_flags)
+static int gpio_stm32_pincfg_to_flags(int otype, int pupd, int mode, int ostate,
+				      gpio_flags_t *out_flags)
 {
-	int flags = 0;
+	gpio_flags_t flags = 0;
 
 	if (mode == LL_GPIO_MODE_OUTPUT) {
 		flags |= GPIO_OUTPUT;
@@ -121,11 +121,17 @@ static int gpio_stm32_pincfg_to_flags(int otype, int pupd, int mode,
 		flags |= GPIO_PULL_DOWN;
 	}
 
+	if (ostate != 0) {
+		flags |= GPIO_OUTPUT_HIGH;
+	} else {
+		flags |= GPIO_OUTPUT_LOW;
+	}
+
 	*out_flags = flags;
 
 	return 0;
 }
-#endif
+#endif /* CONFIG_GPIO_GET_CONFIG */
 
 /**
  * @brief Translate pin to pinval that the LL library needs
@@ -557,16 +563,16 @@ static int gpio_stm32_config(const struct device *dev,
 	return 0;
 }
 
+#ifdef CONFIG_GPIO_GET_CONFIG
 /**
  * @brief Get configuration of pin
  */
-#ifdef CONFIG_GPIO_GET_CONFIG
 static int gpio_stm32_get_config(const struct device *dev,
 				 gpio_pin_t pin, gpio_flags_t *flags)
 {
 	const struct gpio_stm32_config *cfg = dev->config;
 	GPIO_TypeDef *gpio = (GPIO_TypeDef *)cfg->base;
-	unsigned int mode, otype, pupd;
+	unsigned int mode, otype, pupd, ostate;
 	int pin_ll;
 	int err;
 
@@ -579,12 +585,13 @@ static int gpio_stm32_get_config(const struct device *dev,
 	otype = LL_GPIO_GetPinOutputType(gpio, pin_ll);
 	pupd = LL_GPIO_GetPinPull(gpio, pin_ll);
 	mode = LL_GPIO_GetPinMode(gpio, pin_ll);
+	ostate = LL_GPIO_IsOutputPinSet(gpio, pin_ll);
 
-	gpio_stm32_pincfg_to_flags(otype, pupd, mode, flags);
+	gpio_stm32_pincfg_to_flags(otype, pupd, mode, ostate, flags);
 
 	return pm_device_runtime_put(dev);
 }
-#endif
+#endif /* CONFIG_GPIO_GET_CONFIG */
 
 static int gpio_stm32_pin_interrupt_configure(const struct device *dev,
 					      gpio_pin_t pin,
@@ -652,7 +659,7 @@ static const struct gpio_driver_api gpio_stm32_driver = {
 	.pin_configure = gpio_stm32_config,
 #ifdef CONFIG_GPIO_GET_CONFIG
 	.pin_get_config = gpio_stm32_get_config,
-#endif
+#endif /* CONFIG_GPIO_GET_CONFIG */
 	.port_get_raw = gpio_stm32_port_get_raw,
 	.port_set_masked_raw = gpio_stm32_port_set_masked_raw,
 	.port_set_bits_raw = gpio_stm32_port_set_bits_raw,

--- a/drivers/gpio/gpio_stm32.c
+++ b/drivers/gpio/gpio_stm32.c
@@ -95,33 +95,33 @@ static int gpio_stm32_flags_to_conf(gpio_flags_t flags, int *pincfg)
 /**
  * @brief Custom stm32 flags to zephyr
  */
-static int gpio_stm32_pincfg_to_flags(int otype, int pupd, int mode, int ostate,
+static int gpio_stm32_pincfg_to_flags(struct gpio_stm32_pin pin_cfg,
 				      gpio_flags_t *out_flags)
 {
 	gpio_flags_t flags = 0;
 
-	if (mode == LL_GPIO_MODE_OUTPUT) {
+	if (pin_cfg.mode == LL_GPIO_MODE_OUTPUT) {
 		flags |= GPIO_OUTPUT;
-		if (otype == LL_GPIO_OUTPUT_OPENDRAIN) {
+		if (pin_cfg.type == LL_GPIO_OUTPUT_OPENDRAIN) {
 			flags |= GPIO_OPEN_DRAIN;
 		}
-	} else if (mode == LL_GPIO_MODE_INPUT) {
+	} else if (pin_cfg.mode == LL_GPIO_MODE_INPUT) {
 		flags |= GPIO_INPUT;
 #ifdef CONFIG_SOC_SERIES_STM32F1X
-	} else if (mode == LL_GPIO_MODE_FLOATING) {
+	} else if (pin_cfg.mode == LL_GPIO_MODE_FLOATING) {
 		flags |= GPIO_INPUT;
 #endif
 	} else {
 		flags |= GPIO_DISCONNECTED;
 	}
 
-	if (pupd == LL_GPIO_PULL_UP) {
+	if (pin_cfg.pupd == LL_GPIO_PULL_UP) {
 		flags |= GPIO_PULL_UP;
-	} else if (pupd == LL_GPIO_PULL_DOWN) {
+	} else if (pin_cfg.pupd == LL_GPIO_PULL_DOWN) {
 		flags |= GPIO_PULL_DOWN;
 	}
 
-	if (ostate != 0) {
+	if (pin_cfg.out_state != 0) {
 		flags |= GPIO_OUTPUT_HIGH;
 	} else {
 		flags |= GPIO_OUTPUT_LOW;
@@ -572,7 +572,7 @@ static int gpio_stm32_get_config(const struct device *dev,
 {
 	const struct gpio_stm32_config *cfg = dev->config;
 	GPIO_TypeDef *gpio = (GPIO_TypeDef *)cfg->base;
-	unsigned int mode, otype, pupd, ostate;
+	struct gpio_stm32_pin pin_config;
 	int pin_ll;
 	int err;
 
@@ -582,12 +582,12 @@ static int gpio_stm32_get_config(const struct device *dev,
 	}
 
 	pin_ll = stm32_pinval_get(pin);
-	otype = LL_GPIO_GetPinOutputType(gpio, pin_ll);
-	pupd = LL_GPIO_GetPinPull(gpio, pin_ll);
-	mode = LL_GPIO_GetPinMode(gpio, pin_ll);
-	ostate = LL_GPIO_IsOutputPinSet(gpio, pin_ll);
+	pin_config.type = LL_GPIO_GetPinOutputType(gpio, pin_ll);
+	pin_config.pupd = LL_GPIO_GetPinPull(gpio, pin_ll);
+	pin_config.mode = LL_GPIO_GetPinMode(gpio, pin_ll);
+	pin_config.out_state = LL_GPIO_IsOutputPinSet(gpio, pin_ll);
 
-	gpio_stm32_pincfg_to_flags(otype, pupd, mode, ostate, flags);
+	gpio_stm32_pincfg_to_flags(pin_config, flags);
 
 	return pm_device_runtime_put(dev);
 }

--- a/drivers/gpio/gpio_stm32.h
+++ b/drivers/gpio/gpio_stm32.h
@@ -206,7 +206,7 @@
 #define STM32_PINCFG_FLOATING           STM32_PUPDR_NO_PULL
 #endif /* CONFIG_SOC_SERIES_STM32F1X */
 
-#ifdef CONFIG_GPIO_GET_CONFIG
+#if defined(CONFIG_GPIO_GET_CONFIG) && !defined(CONFIG_SOC_SERIES_STM32F1X)
 /**
  * @brief structure of a GPIO pin (stm32 LL values) use to get the configuration
  */

--- a/drivers/gpio/gpio_stm32.h
+++ b/drivers/gpio/gpio_stm32.h
@@ -206,6 +206,18 @@
 #define STM32_PINCFG_FLOATING           STM32_PUPDR_NO_PULL
 #endif /* CONFIG_SOC_SERIES_STM32F1X */
 
+#ifdef CONFIG_GPIO_GET_CONFIG
+/**
+ * @brief structure of a GPIO pin (stm32 LL values) use to get the configuration
+ */
+struct gpio_stm32_pin {
+	unsigned int type; /* LL_GPIO_OUTPUT_PUSHPULL or LL_GPIO_OUTPUT_OPENDRAIN */
+	unsigned int pupd; /* LL_GPIO_PULL_NO or LL_GPIO_PULL_UP or LL_GPIO_PULL_DOWN */
+	unsigned int mode; /* LL_GPIO_MODE_INPUT or LL_GPIO_MODE_OUTPUT or other */
+	unsigned int out_state; /* 1 (high level) or 0 (low level) */
+};
+#endif /* CONFIG_GPIO_GET_CONFIG */
+
 /**
  * @brief configuration of GPIO device
  */

--- a/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c
+++ b/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c
@@ -644,7 +644,7 @@ static int pin_get_config(void)
 	zassert_equal(rc, 0, "pin configure failed");
 
 	rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
-	if (rc == -ENOTSUP) {
+	if (rc == -ENOSYS) {
 		return TC_PASS;
 	}
 

